src/internal/group_api/mod.rs

```diff
@@ -453,108 +450,95 @@ fn collect_admin_and_member_info<CR: rand::CryptoRng + rand::RngCore>(
-                            admins,
-                            members,
-                            user_ids_and_keys,
-                        )?;
-
-                        let group_members: Vec<requests::GroupMember> = member_info
-                            .into_iter()
-                            .map(|(member_id, member_pub_key, member_trans_key)| {
-                                requests::GroupMember {
-                                    user_id: member_id,
-                                    transform_key: member_trans_key.into(),
-                                    user_master_public_key: member_pub_key.into(),
-                                }
-                            })
-                            .collect();
-                        let maybe_group_members = if group_members.is_empty() {
-                            None
-                        } else {
-                            Some(group_members)
-                        };
-
-                        let group_admins: Vec<requests::GroupAdmin> = admin_info
-                            .into_iter()
-                            .map(|(admin_id, admin_pub_key)| {
-                                let encrypted_group_key = recrypt.encrypt(
-                                    &plaintext,
-                                    &admin_pub_key.clone().into(),
-                                    &auth.signing_private_key().into(),
-                                );
-                                encrypted_group_key
-                                    .map_err(|e| e.into())
-                                    .and_then(EncryptedOnceValue::try_from)
-                                    .map(|enc_msg| requests::GroupAdmin {
-                                        encrypted_msg: enc_msg,
-                                        user: requests::User {
-                                            user_id: admin_id,
-                                            user_master_public_key: admin_pub_key.into(),
-                                        },
-                                    })
-                            })
-                            .collect::<Result<Vec<_>, _>>()?;
-
-                        Ok((group_pub_key, maybe_group_members, group_admins))
-                    })
-                    .into_future()
-            }
-        })
-        .and_then(move |(group_pub_key, member_vec, admin_vec)| {
-            requests::group_create::group_create(
-                &auth,
-                group_id,
-                name,
-                group_pub_key,
-                owner,
-                admin_vec,
-                member_vec,
-                needs_rotation,
-            )
+) -> Result<GroupCreateResult, IronOxideErr> {
+    let user_ids_and_keys = user_api::user_key_list(auth, users_to_lookup).await?;
+    // this will occur when one of the UserIds cannot be found
+    let user_ids_and_keys = check_user_mismatch(users_to_lookup, user_ids_and_keys)?;
```

Comment:
I like this new layout much better. But what's the reason for this reassignment?

Reply:
The check_user_mismatch takes a value and will either produce an error message from it or return it. It allows us to avoid cloning the entire map.